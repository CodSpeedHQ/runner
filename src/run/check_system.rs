use std::process::Command;

use lazy_static::lazy_static;
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use sysinfo::{CpuRefreshKind, MemoryRefreshKind, RefreshKind, System};

use crate::prelude::*;

fn get_user() -> Result<String> {
    let user_output = Command::new("whoami")
        .output()
        .map_err(|_| anyhow!("Failed to get user info"))?;
    if !user_output.status.success() {
        bail!("Failed to get user info");
    }
    let output_str =
        String::from_utf8(user_output.stdout).map_err(|_| anyhow!("Failed to parse user info"))?;
    Ok(output_str.trim().to_string())
}

#[derive(Eq, PartialEq, Hash, Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct SystemInfo {
    pub os: String,
    pub os_version: String,
    pub arch: String,
    pub host: String,
    pub user: String,
    pub cpu_brand: String,
    pub cpu_name: String,
    pub cpu_vendor_id: String,
    pub cpu_cores: usize,
    pub total_memory_gb: u64,
}

#[cfg(test)]
impl SystemInfo {
    pub fn test() -> Self {
        SystemInfo {
            os: "ubuntu".to_string(),
            os_version: "20.04".to_string(),
            arch: "x86_64".to_string(),
            host: "host".to_string(),
            user: "user".to_string(),
            cpu_brand: "Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz".to_string(),
            cpu_name: "cpu0".to_string(),
            cpu_vendor_id: "GenuineIntel".to_string(),
            cpu_cores: 2,
            total_memory_gb: 8,
        }
    }
}

impl SystemInfo {
    pub fn new() -> Result<Self> {
        let os = System::distribution_id();
        let os_version = System::os_version().ok_or(anyhow!("Failed to get OS version"))?;
        let arch = System::cpu_arch();
        let user = get_user()?;
        let host = System::host_name().ok_or(anyhow!("Failed to get host name"))?;

        let s = System::new_with_specifics(
            RefreshKind::nothing()
                .with_cpu(CpuRefreshKind::everything())
                .with_memory(MemoryRefreshKind::everything()),
        );
        let cpu_cores = s
            .physical_core_count()
            .ok_or(anyhow!("Failed to get CPU core count"))?;
        let total_memory_gb = s.total_memory().div_ceil(1024_u64.pow(3));

        // take the first CPU to get the brand, name and vendor id
        let cpu = s
            .cpus()
            .iter()
            .next()
            .ok_or(anyhow!("Failed to get CPU info"))?;
        let cpu_brand = cpu.brand().to_string();
        let cpu_name = cpu.name().to_string();
        let cpu_vendor_id = cpu.vendor_id().to_string();

        Ok(SystemInfo {
            os,
            os_version,
            arch,
            host,
            user,
            cpu_brand,
            cpu_name,
            cpu_vendor_id,
            cpu_cores,
            total_memory_gb,
        })
    }
}

lazy_static! {
    static ref SUPPORTED_SYSTEMS: HashSet<(&'static str, &'static str, &'static str)> = {
        HashSet::from([
            ("ubuntu", "22.04", "x86_64"),
            ("ubuntu", "24.04", "x86_64"),
            ("ubuntu", "22.04", "aarch64"),
            ("ubuntu", "24.04", "aarch64"),
            ("debian", "12", "x86_64"),
            ("debian", "12", "aarch64"),
        ])
    };
}

/// Checks if the provided system info is supported
///
/// Supported systems:
/// - Ubuntu 20.04 x86_64
/// - Ubuntu 22.04 x86_64 and aarch64
/// - Debian 11 x86_64
/// - Debian 12 x86_64
pub fn check_system(system_info: &SystemInfo) -> Result<()> {
    debug!("System info: {:#?}", system_info);

    let system_tuple = (
        system_info.os.as_str(),
        system_info.os_version.as_str(),
        system_info.arch.as_str(),
    );

    if SUPPORTED_SYSTEMS.contains(&system_tuple) {
        return Ok(());
    }

    match system_info.arch.as_str() {
        "x86_64" | "aarch64" => {
            warn!(
                "Unofficially supported system: {} {}. Continuing with best effort support.",
                system_info.os, system_info.os_version
            );
            return Ok(());
        }
        _ => {}
    }

    bail!(
        "Unsupported system: {} {}",
        system_info.os,
        system_info.os_version
    );
}
