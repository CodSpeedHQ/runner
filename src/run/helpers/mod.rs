mod download_file;
mod find_repository_root;
mod get_env_var;
mod parse_git_remote;

pub use download_file::download_file;
pub use find_repository_root::find_repository_root;
pub use get_env_var::get_env_variable;
pub use parse_git_remote::*;
