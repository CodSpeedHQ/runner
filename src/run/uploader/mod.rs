mod interfaces;
mod upload;
mod upload_metadata;

pub use interfaces::*;
pub use upload::upload;
