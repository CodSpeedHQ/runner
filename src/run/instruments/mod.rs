use std::collections::HashSet;

use log::warn;
use serde::{Deserialize, Serialize};

use crate::prelude::*;
use crate::run::RunArgs;

pub mod mongo_tracer;

#[derive(Debug, PartialEq, Eq)]
pub struct MongoDBConfig {
    pub uri_env_name: Option<String>,
}

#[derive(Debug, PartialEq, Eq)]
pub struct Instruments {
    pub mongodb: Option<MongoDBConfig>,
}

#[derive(Deserialize, Serialize, Debug, Clone, PartialEq, Eq, Hash)]
pub enum InstrumentNames {
    MongoDB,
}

impl Instruments {
    pub fn is_mongodb_enabled(&self) -> bool {
        self.mongodb.is_some()
    }

    pub fn get_active_instrument_names(&self) -> Vec<InstrumentNames> {
        let mut names = vec![];

        if self.is_mongodb_enabled() {
            names.push(InstrumentNames::MongoDB);
        }

        names
    }
}

impl TryFrom<&RunArgs> for Instruments {
    type Error = Error;
    fn try_from(args: &RunArgs) -> Result<Self> {
        let mut validated_instrument_names: HashSet<InstrumentNames> = HashSet::new();

        for instrument_name in &args.instruments {
            match instrument_name.as_str() {
                "mongodb" => validated_instrument_names.insert(InstrumentNames::MongoDB),
                _ => bail!("Invalid instrument name: {}", instrument_name),
            };
        }

        let mongodb = if validated_instrument_names.contains(&InstrumentNames::MongoDB) {
            Some(MongoDBConfig {
                uri_env_name: args.mongo_uri_env_name.clone(),
            })
        } else if args.mongo_uri_env_name.is_some() {
            warn!("The MongoDB instrument is disabled but a MongoDB URI environment variable name was provided, ignoring it");
            None
        } else {
            None
        };

        Ok(Self { mongodb })
    }
}

#[cfg(test)]
impl Instruments {
    /// Constructs a new `Instruments` with default values for testing purposes
    pub fn test() -> Self {
        Self {
            mongodb: Some(MongoDBConfig {
                uri_env_name: Some("MONGODB_URI".into()),
            }),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_from_args_empty() {
        let instruments = Instruments::try_from(&RunArgs::test()).unwrap();
        assert!(instruments.mongodb.is_none());
    }

    #[test]
    fn test_from_args() {
        let args = RunArgs {
            instruments: vec!["mongodb".into()],
            mongo_uri_env_name: Some("MONGODB_URI".into()),
            ..RunArgs::test()
        };
        let instruments = Instruments::try_from(&args).unwrap();
        assert_eq!(
            instruments.mongodb,
            Some(MongoDBConfig {
                uri_env_name: Some("MONGODB_URI".into())
            })
        );
        assert!(instruments.is_mongodb_enabled());
    }

    #[test]
    fn test_from_args_mongodb_disabled() {
        let args = RunArgs {
            instruments: vec![],
            mongo_uri_env_name: Some("MONGODB_URI".into()),
            ..RunArgs::test()
        };
        let instruments = Instruments::try_from(&args).unwrap();
        assert_eq!(instruments.mongodb, None);
        assert!(!instruments.is_mongodb_enabled());
    }

    #[test]
    fn test_from_args_unknown_instrument_value() {
        let args = RunArgs {
            instruments: vec!["unknown".into()],
            mongo_uri_env_name: Some("MONGODB_URI".into()),
            ..RunArgs::test()
        };
        let instruments = Instruments::try_from(&args);
        assert!(instruments.is_err());
        assert_eq!(
            instruments.unwrap_err().to_string(),
            "Invalid instrument name: unknown"
        );
    }
}
