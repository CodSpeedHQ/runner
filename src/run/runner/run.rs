use crate::prelude::*;
use crate::run::{
    check_system::SystemInfo, config::Config, instruments::mongo_tracer::MongoTracer,
};

use std::path::PathBuf;

use super::{
    helpers::{perf_maps::harvest_perf_maps, profile_folder::create_profile_folder},
    setup::setup,
    valgrind,
};

pub struct RunData {
    pub profile_folder: PathBuf,
}

pub async fn run(config: &Config, system_info: &SystemInfo) -> Result<RunData> {
    if !config.skip_setup {
        start_group!("Preparing the environment");
        setup(system_info, config).await?;
        end_group!();
    }
    //TODO: add valgrind version check
    start_opened_group!("Running the benchmarks");
    let profile_folder = create_profile_folder()?;
    let mongo_tracer = if let Some(mongodb_config) = &config.instruments.mongodb {
        let mut mongo_tracer = MongoTracer::try_from(&profile_folder, mongodb_config)?;
        mongo_tracer.start().await?;
        Some(mongo_tracer)
    } else {
        None
    };
    valgrind::measure(config, &profile_folder, &mongo_tracer)?;
    harvest_perf_maps(&profile_folder)?;
    if let Some(mut mongo_tracer) = mongo_tracer {
        mongo_tracer.stop().await?;
    }
    end_group!();
    Ok(RunData { profile_folder })
}
