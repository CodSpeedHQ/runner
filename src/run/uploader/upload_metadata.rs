use serde_json::json;

use super::UploadMetadata;

impl UploadMetadata {
    pub fn get_hash(&self) -> String {
        let upload_metadata_string = json!(&self).to_string();
        sha256::digest(upload_metadata_string)
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_json_snapshot;

    use crate::run::{
        check_system::SystemInfo,
        ci_provider::interfaces::{GhData, ProviderMetadata, RunEvent, Sender},
        instruments::InstrumentName,
        runner::ExecutorName,
        uploader::{Runner, UploadMetadata},
    };

    #[test]
    fn test_get_metadata_hash() {
        let upload_metadata = UploadMetadata {
            version: Some(4),
            tokenless: true,
            profile_md5: "jp/k05RKuqP3ERQuIIvx4Q==".into(),
            runner: Runner {
                name: "codspeed-runner".into(),
                version: "2.1.0".into(),
                instruments: vec![InstrumentName::MongoDB],
                executor: ExecutorName::Valgrind,
                system_info: SystemInfo::test(),
            },
            platform: "github-actions".into(),
            commit_hash: "5bd77cb0da72bef094893ed45fb793ff16ecfbe3".into(),
            provider_metadata: ProviderMetadata {
                ref_: "refs/pull/29/merge".into(),
                head_ref: Some("chore/native-action-runner".into()),
                base_ref: Some("main".into()),
                owner: "CodSpeedHQ".into(),
                repository: "codspeed-node".into(),
                event: RunEvent::PullRequest,
                gh_data: Some(GhData {
                    run_id: "7044765741".into(),
                    job: "codspeed".into(),
                    sender: Some(Sender {
                        id: 19605940,
                        login: "adriencaccia".into(),
                    }),
                }),
                repository_root_path: "/home/runner/work/codspeed-node/codspeed-node/".into(),
            },
        };

        let hash = upload_metadata.get_hash();
        assert_eq!(
            hash,
            "b85b85835e4fc5d222e35fa762353dfcfecbed157ee866323bffa4c11151f66d"
        );
        assert_json_snapshot!(upload_metadata);
    }
}
