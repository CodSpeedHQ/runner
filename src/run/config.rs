use crate::prelude::*;
use crate::run::instruments::Instruments;
use url::Url;

use crate::run::RunArgs;

#[derive(Debug)]
pub struct Config {
    pub upload_url: Url,
    pub token: Option<String>,
    pub working_directory: Option<String>,
    pub command: String,

    pub instruments: Instruments,

    pub skip_upload: bool,
    pub skip_setup: bool,
}

#[cfg(test)]
impl Config {
    /// Constructs a new `Config` with default values for testing purposes
    pub fn test() -> Self {
        Self {
            upload_url: Url::parse(DEFAULT_UPLOAD_URL).unwrap(),
            token: None,
            working_directory: None,
            command: "".into(),
            instruments: Instruments::test(),
            skip_upload: false,
            skip_setup: false,
        }
    }
}

const DEFAULT_UPLOAD_URL: &str = "https://api.codspeed.io/upload";

impl TryFrom<RunArgs> for Config {
    type Error = Error;
    fn try_from(args: RunArgs) -> Result<Self> {
        let instruments = Instruments::try_from(&args)?;
        let raw_upload_url = args.upload_url.unwrap_or_else(|| DEFAULT_UPLOAD_URL.into());
        let upload_url = Url::parse(&raw_upload_url)
            .map_err(|e| anyhow!("Invalid upload URL: {}, {}", raw_upload_url, e))?;
        Ok(Self {
            upload_url,
            token: args.token,
            working_directory: args.working_directory,
            instruments,
            command: args.command.join(" "),
            skip_upload: args.skip_upload,
            skip_setup: args.skip_setup,
        })
    }
}

#[cfg(test)]
mod tests {
    use crate::run::instruments::MongoDBConfig;

    use super::*;

    #[test]
    fn test_try_from_env_empty() {
        let config = Config::try_from(RunArgs {
            upload_url: None,
            token: None,
            working_directory: None,
            instruments: vec![],
            mongo_uri_env_name: None,
            skip_upload: false,
            skip_setup: false,
            command: vec!["cargo".into(), "codspeed".into(), "bench".into()],
        })
        .unwrap();
        assert_eq!(config.upload_url, Url::parse(DEFAULT_UPLOAD_URL).unwrap());
        assert_eq!(config.token, None);
        assert_eq!(config.working_directory, None);
        assert_eq!(config.instruments, Instruments { mongodb: None });
        assert!(!config.skip_upload);
        assert!(!config.skip_setup);
        assert_eq!(config.command, "cargo codspeed bench");
    }

    #[test]
    fn test_try_from_args() {
        let config = Config::try_from(RunArgs {
            upload_url: Some("https://example.com/upload".into()),
            token: Some("token".into()),
            working_directory: Some("/tmp".into()),
            instruments: vec!["mongodb".into()],
            mongo_uri_env_name: Some("MONGODB_URI".into()),
            skip_upload: true,
            skip_setup: true,
            command: vec!["cargo".into(), "codspeed".into(), "bench".into()],
        })
        .unwrap();

        assert_eq!(
            config.upload_url,
            Url::parse("https://example.com/upload").unwrap()
        );
        assert_eq!(config.token, Some("token".into()));
        assert_eq!(config.working_directory, Some("/tmp".into()));
        assert_eq!(
            config.instruments,
            Instruments {
                mongodb: Some(MongoDBConfig {
                    uri_env_name: Some("MONGODB_URI".into())
                })
            }
        );
        assert!(config.skip_upload);
        assert!(config.skip_setup);
        assert_eq!(config.command, "cargo codspeed bench");
    }
}
