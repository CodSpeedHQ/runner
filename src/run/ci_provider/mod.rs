pub mod interfaces;
pub mod logger;
mod provider;

use buildkite::BuildkiteProvider;
use github_actions::GitHubActionsProvider;
use local::LocalProvider;
use provider::CIProviderDetector;

use crate::prelude::*;
use crate::run::config::Config;

pub use self::provider::CIProvider;

// Provider implementations
mod buildkite;
mod github_actions;
mod local;

pub fn get_provider(config: &Config) -> Result<Box<dyn CIProvider>> {
    if BuildkiteProvider::detect() {
        let provider = BuildkiteProvider::try_from(config)?;
        return Ok(Box::new(provider));
    }

    if GitHubActionsProvider::detect() {
        let provider = GitHubActionsProvider::try_from(config)?;
        return Ok(Box::new(provider));
    }

    if LocalProvider::detect() {
        let provider = LocalProvider::try_from(config)?;
        return Ok(Box::new(provider));
    }

    bail!("No CI provider detected")
}
