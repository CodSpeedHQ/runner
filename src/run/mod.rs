use crate::api_client::CodSpeedAPIClient;
use crate::config::CodSpeedConfig;
use crate::prelude::*;
use crate::run::{config::Config, logger::Logger};
use crate::VERSION;
use check_system::SystemInfo;
use clap::Args;
use instruments::mongo_tracer::MongoTracer;
use runner::get_run_data;

mod check_system;
pub mod ci_provider;
mod helpers;
mod instruments;
mod poll_results;
mod runner;
mod uploader;

pub mod config;
pub mod logger;

fn show_banner() {
    let banner = format!(
        r#"
   ______            __ _____                         __
  / ____/____   ____/ // ___/ ____   ___   ___   ____/ /
 / /    / __ \ / __  / \__ \ / __ \ / _ \ / _ \ / __  /
/ /___ / /_/ // /_/ / ___/ // /_/ //  __//  __// /_/ /
\____/ \____/ \__,_/ /____// .___/ \___/ \___/ \__,_/
  https://codspeed.io     /_/          runner v{}
"#,
        VERSION
    );
    println!("{}", banner);
    debug!("codspeed v{}", VERSION);
}

#[derive(Args, Debug)]
pub struct RunArgs {
    /// The upload URL to use for uploading the results, useful for on-premises installations
    #[arg(long)]
    pub upload_url: Option<String>,

    /// The token to use for uploading the results,
    #[arg(long, env = "CODSPEED_TOKEN")]
    pub token: Option<String>,

    /// The directory where the command will be executed.
    #[arg(long)]
    pub working_directory: Option<String>,

    /// Comma-separated list of instruments to enable. Possible values: mongodb.
    #[arg(long, value_delimiter = ',')]
    pub instruments: Vec<String>,

    /// The name of the environment variable that contains the MongoDB URI to patch.
    /// If not provided, user will have to provide it dynamically through a CodSpeed integration.
    ///
    /// Only used if the `mongodb` instrument is enabled.
    #[arg(long)]
    pub mongo_uri_env_name: Option<String>,

    /// Only for debugging purposes, skips the upload of the results
    #[arg(
        long,
        default_value = "false",
        hide = true,
        env = "CODSPEED_SKIP_UPLOAD"
    )]
    pub skip_upload: bool,

    /// Only for debugging purposes, skips the setup of the runner
    #[arg(long, default_value = "false", hide = true)]
    pub skip_setup: bool,

    /// The bench command to run
    pub command: Vec<String>,
}

#[cfg(test)]
impl RunArgs {
    /// Constructs a new `RunArgs` with default values for testing purposes
    pub fn test() -> Self {
        Self {
            upload_url: None,
            token: None,
            working_directory: None,
            instruments: vec![],
            mongo_uri_env_name: None,
            skip_upload: false,
            skip_setup: false,
            command: vec![],
        }
    }
}

pub async fn run(args: RunArgs, api_client: &CodSpeedAPIClient) -> Result<()> {
    let mut config = Config::try_from(args)?;
    let provider = ci_provider::get_provider(&config)?;
    let codspeed_config = CodSpeedConfig::load()?;
    let logger = Logger::new(&provider)?;

    if provider.get_provider_slug() != "local" {
        show_banner();
    }
    debug!("config: {:#?}", config);

    if provider.get_provider_slug() == "local" {
        if codspeed_config.auth.token.is_none() {
            bail!("You have to authenticate the CLI first. Run `codspeed auth login`.");
        }
        debug!("Using the token from the CodSpeed configuration file");
        config.set_token(codspeed_config.auth.token.clone());
    }

    let system_info = SystemInfo::new()?;
    check_system::check_system(&system_info)?;

    let mode = runner::get_mode()?;
    let executor = runner::get_executor_from_mode(mode);

    let run_data = get_run_data()?;

    if !config.skip_setup {
        start_group!("Preparing the environment");
        executor.setup(&config, &system_info, &run_data).await?;
        end_group!();
    }

    start_opened_group!("Running the benchmarks");

    // TODO: refactor and move directly in the Instruments struct as a `start` method
    let mongo_tracer = if let Some(mongodb_config) = &config.instruments.mongodb {
        let mut mongo_tracer = MongoTracer::try_from(&run_data.profile_folder, mongodb_config)?;
        mongo_tracer.start().await?;
        Some(mongo_tracer)
    } else {
        None
    };

    executor
        .run(&config, &system_info, &run_data, &mongo_tracer)
        .await?;

    // TODO: refactor and move directly in the Instruments struct as a `stop` method
    if let Some(mut mongo_tracer) = mongo_tracer {
        mongo_tracer.stop().await?;
    }

    executor.teardown(&config, &system_info, &run_data).await?;

    end_group!();

    if !config.skip_upload {
        start_group!("Uploading performance data");
        logger.persist_log_to_profile_folder(&run_data)?;
        let upload_result =
            uploader::upload(&config, &system_info, &provider, &run_data, executor.name()).await?;
        end_group!();

        if provider.get_provider_slug() == "local" {
            start_group!("Fetching the results");
            poll_results::poll_results(api_client, &provider, upload_result.run_id).await?;
            end_group!();
        }
    }

    Ok(())
}
