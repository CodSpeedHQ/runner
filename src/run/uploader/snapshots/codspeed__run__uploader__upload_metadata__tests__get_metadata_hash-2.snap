---
source: src/run/uploader/upload_metadata.rs
expression: upload_metadata
---
{
  "repositoryProvider": "GITHUB",
  "version": 6,
  "tokenless": true,
  "profileMd5": "jp/k05RKuqP3ERQuIIvx4Q==",
  "runner": {
    "name": "codspeed-runner",
    "version": "2.1.0",
    "instruments": [
      "MongoDB"
    ],
    "executor": "valgrind",
    "os": "ubuntu",
    "osVersion": "20.04",
    "arch": "x86_64",
    "host": "host",
    "user": "user",
    "cpuBrand": "Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz",
    "cpuName": "cpu0",
    "cpuVendorId": "GenuineIntel",
    "cpuCores": 2,
    "totalMemoryGb": 8
  },
  "runEnvironment": "GITHUB_ACTIONS",
  "runPart": {
    "runId": "7044765741",
    "runPartId": "benchmarks_3.2.2",
    "jobName": "codspeed",
    "metadata": {
      "anotherKey": "anotherValue",
      "someKey": "someValue"
    }
  },
  "commitHash": "5bd77cb0da72bef094893ed45fb793ff16ecfbe3",
  "ref": "refs/pull/29/merge",
  "headRef": "chore/native-action-runner",
  "baseRef": "main",
  "owner": "CodSpeedHQ",
  "repository": "codspeed-node",
  "event": "pull_request",
  "sender": {
    "id": "19605940",
    "login": "adriencaccia"
  },
  "ghData": {
    "runId": "7044765741",
    "job": "codspeed"
  },
  "glData": null,
  "repositoryRootPath": "/home/runner/work/codspeed-node/codspeed-node/"
}
